File: server/src/main/java/com/textchat/textmessages/TextMessagesController.java
package com.textchat.textmessages;

import com.textchat.persistence.contacts.Contact;
import com.textchat.persistence.contacts.ContactRepository;
import com.textchat.persistence.textmessages.TextMessage;
import com.textchat.persistence.textmessages.TextMessageRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletRequest;
import java.util.List;
import java.util.stream.Stream;

import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static java.util.stream.Collectors.toList;

@RestController
public class TextMessagesController {
    private TextMessageRepository textMessageRepository;
    private ContactRepository contactRepository;
    private SimpMessagingTemplate simpMessagingTemplate;
    private TextMessageService textMessageService;

    @Autowired
    public TextMessagesController(
            TextMessageRepository textMessageRepository,
            ContactRepository contactRepository,
            SimpMessagingTemplate simpMessagingTemplate,
            TextMessageService textMessageService
    ) {
        this.textMessageRepository = textMessageRepository;
        this.contactRepository = contactRepository;
        this.simpMessagingTemplate = simpMessagingTemplate;
        this.textMessageService = textMessageService;
    }

    @GetMapping("/text-messages")
    public AugmentedTextMessageResponse list() {
        List<TextMessage> latestThreads = textMessageRepository
                .findLatestThreads();

        List<TextMessageResponse> textMessageResponses = latestThreads
                .stream()
                .map(textMessage -> new TextMessageResponse(textMessage, false)) // TODO
                .collect(toList());

        List<ContactResponse> contactResponses = latestThreads
                .stream()
                .flatMap(textMessage ->
                        Stream.of(
                                new ContactResponse(textMessage.getToContact()),
                                new ContactResponse(textMessage.getFromContact())
                        )
                )
                .collect(toList());

        return new AugmentedTextMessageResponse(textMessageResponses, contactResponses);
    }

    @GetMapping("/contacts/{contactId}/text-messages")
    public List<TextMessageResponse> listTextMessagesForContact(@PathVariable Long contactId) {
        Contact contact = contactRepository.findOne(contactId);

        return textMessageRepository
                .findAllByToContactOrFromContactOrderByCreatedAtDesc(contact, contact)
                .stream()
                .map(textMessage -> new TextMessageResponse(textMessage, false)) // TODO
                .collect(toList());
    }

    @PostMapping("/contacts/{contactId}/text-messages")
    public TextMessageResponse createForContact(
            @RequestBody SendMessageRequest sendMessageRequest,
            @PathVariable Long contactId
    ) {

        TextMessage textMessage = textMessageService.send(
                contactRepository.findOne(contactId),
                sendMessageRequest.getBody()
        );

        TextMessageResponse textMessageResponse = new TextMessageResponse(textMessage, false);

        simpMessagingTemplate.convertAndSend("/text-messages",
                new AugmentedTextMessageResponse(
                        singletonList(textMessageResponse),
                        asList(
                                new ContactResponse(textMessage.getToContact()),
                                new ContactResponse(textMessage.getFromContact())
                        )
                )
        );

        return textMessageResponse; // TODO
    }

    // TODO: add some security here so random people can't hit this endpoint
    @PostMapping("/receive-sms")
    public void receiveTextMessage(HttpServletRequest httpServletRequest) {
        String body = httpServletRequest.getParameter("Body");
        String to = httpServletRequest.getParameter("To");
        String from = httpServletRequest.getParameter("From");

        TextMessage textMessage = textMessageService.recordReceipt(from, to, body);

        simpMessagingTemplate.convertAndSend("/text-messages", new AugmentedTextMessageResponse(
                singletonList(new TextMessageResponse(textMessage, false)),
                asList(
                        new ContactResponse(textMessage.getToContact()),
                        new ContactResponse(textMessage.getFromContact()))
                )
        );
    }
}

File: ruby-server/app/controllers/api/text_messages_controller.rb
class Api::TextMessagesController < ApplicationController
  include ChannelHelper

  skip_before_action :authenticate_user!, only: :receive

  def index
    if params[:contact_id].present?
      contact = Contact.where(account: current_account).find(params[:contact_id])
      @text_messages = TextMessage
                           .where(account: current_account)
                           .for_to_or_from_contact(contact)
                           .order(created_at: :desc)
      return render :contact_index
    end

    @text_messages = TextMessage
                         .includes(:to_contact, :from_contact)
                         .latest_threads(current_account)
    @contacts = @text_messages.map(&:to_contact) + @text_messages.map(&:from_contact)
  end

  def create
    to_contact = Contact.find(params[:contact_id])
    @text_message = TextMessage.new(
        account: current_account,
        body: params[:body],
        to_contact: to_contact,
        from_contact: current_contact
    )
    TextMessageService.send_text_message(@text_message)
    broadcast_text_messages [@text_message]
  end

  # TODO: add some security here so random people can't hit this endpoint
  def receive
    text_message = TextMessageService.record_receipt(params["msisdn"], params["to"], params["text"])
    broadcast_text_messages [text_message]
  end
end